neurobooth_os/iout/metadator.py

```diff
@@ -404,3 +432,121 @@ def get_device_kwargs_by_task(collection_id, conn) -> OrderedDict:
     return tasks_kwarg
 
 
+def read_sensors() -> Dict[str, SensorArgs]:
+    """Return dictionary of sensor_id to SensorArgs for all yaml sensor parameter files."""
+
+    directory: str = get_cfg_path("sensors")
+    sens_dict = {}
+    for file in os.listdir(directory):
+        file_name = os.fsdecode(file).split(".")[0]
+        param_dict: Dict[str:Any] = stim_param_reader.get_param_dictionary(file, 'sensors')
+        param_parser: str = param_dict['arg_parser']
+        parser_func = str_fileid_to_eval(param_parser)
+        args: SensorArgs = parser_func(**param_dict)
+        sens_dict[file_name] = args
+    return sens_dict
+
+
+def read_devices() -> Dict[str, DeviceArgs]:
+    """Return dictionary of device_id to DeviceArgs for all yaml device parameter files."""
+
+    directory: str = get_cfg_path("devices")
+    sens_dict = {}
+    for file in os.listdir(directory):
+        file_name = os.fsdecode(file).split(".")[0]
+        param_dict: Dict[str:Any] = stim_param_reader.get_param_dictionary(file, 'devices')
+        param_parser: str = param_dict['arg_parser']
+        parser_func = str_fileid_to_eval(param_parser)
+        args: DeviceArgs = parser_func(**param_dict)
+        sens_dict[file_name] = args
+    return sens_dict
+
+
+def read_instructions() -> Dict[str, InstructionArgs]:
+    """Return dictionary of instruction_id to InstructionArgs for all yaml instruction parameter files."""
+
+    directory: str = get_cfg_path("instructions")
+    sens_dict = {}
+    for file in os.listdir(directory):
+        file_name = os.fsdecode(file).split(".")[0]
+        param_dict: Dict[str:Any] = stim_param_reader.get_param_dictionary(file, 'instructions')
+        param_parser: str = param_dict['arg_parser']
+        parser_func = str_fileid_to_eval(param_parser)
+        args: InstructionArgs = parser_func(**param_dict)
```

Comment:
Maybe this is too opaque, but we could write a `dynamic_parse()` function. E,g.:
```
def dynamic_parse(file: str, param_type: str, model_class: ClassVar) -> BaseModel:
    param_dict: Dict[str:Any] = stim_param_reader.get_param_dictionary(file, param_type)
    param_parser: str = param_dict['arg_parser']
    parser_func = str_fileid_to_eval(param_parser)
    return parser_func(**param_dict)
```

Maybe be able to factor out some of the surrounding code as well into another generic method.

Reply:
Ok.